### PR TITLE
Enable Traces by default in Samples Project (related to #775)

### LINKED ISSUE #775: [FEATURE] Enable Traces by default in Samples Project
**Describe the feature**  
Enable Traces by default in Samples Project. Provide a Documentation if possible how this configuration works. Change LogFile path, enable  and disable it.

**Explain why this feature is needed**  
As Tester, I will run (debug) the sample tests and be able to know  any time which Action is performing EasyRepro in the Output Window (Debug) in Visual Studio.
image

As Tester, I will run all my tests and after the conclusion, I will be able to analyze step by step, which actions have performed my tests. 
image

As Tester, if I run my tests in a DevOps Pipeline, I will be able to see in the test results, step by step, which actions have performed my tests.
image

As Tester, I will have this trace analyse possibilities with a minimum effort.

**Additional context**  
I want to create my own project using the samples project as template having included the traces configuration.

### PR DESCRIPTION
**Pull request by AngelRodriguez8008**
### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Other (updates to documentation, formatting, etc.)

### Description
- Add traces configuration in sample project.
- Add helper classes related to traces.
- Add Trace basic infos about the related test in the UCI TestsBase class.
TODO: The configuration is simple. I add some comments, but create proper Documentation is highly recommended to show the different scenarios; for example: Send Traces to a Text File, use with NUnit, configure in DevOps, etc.

### Issues addressed
See #775 [FEATURE] Enable Traces by default in Samples Project

### All submissions:
- [x] My code follows the code style of this project.
- [x] Do existing samples that are effected by this change still run?
- [x] I have added samples for new functionality. 
- [x] I raise detailed error messages when possible.
- [x] My code does not rely on labels that have the option to be hidden.

### Which browsers was this tested on?

- [x] Chrome
- [ ] Firefox
- [ ] IE
- [ ] Edge

Tests Results on Build Pipeline:
image

Attached Traces (download is allowed):
image

[Standard_Console_Output (example Lookup not found).log](https://github.com/microsoft/EasyRepro/files/4240438/Standard_Console_Output.example.Lookup.not.found.log)

[Standard_Console_Output (example Login Fail).log](https://github.com/microsoft/EasyRepro/files/4240436/Standard_Console_Output.example.Login.Fail.log)

**NOTE**: I found a problem with the standard "System.Diagnostics.ConsoleTraceListener". Using the MSTest just the first test print the traces to the Standard Output, to fix this a custom TraceListener 
 implementation is required. 
See more information about the error & the workaround [here](https://stackoverflow.com/questions/36327531/why-are-messages-sent-to-trace-source-missing-from-all-but-the-first-unit-test).
For NUnit is the custom TraceListener not required, a respective explanatory comment is added in app.config.

### CODE REVIEW

File: Microsoft.Dynamics365.UIAutomation.Sample/UCI/Controls/Grid.cs [hunk 1]
```diff
@@ -3,20 +3,12 @@
...
 {
     [TestClass]
     public class Grid : TestsBase
     {
-
-        private readonly SecureString _username = System.Configuration.ConfigurationManager.AppSettings["OnlineUsername"].ToSecureString();
-        private readonly SecureString _password = System.Configuration.ConfigurationManager.AppSettings["OnlinePassword"].ToSecureString();
-        private readonly Uri _xrmUri = new Uri(System.Configuration.ConfigurationManager.AppSettings["OnlineCrmUrl"].ToString());
-
         [TestMethod]
```
Comment (AngelRodriguez8008) [inline Microsoft.Dynamics365.UIAutomation.Sample/UCI/Controls/Grid.cs:12]: Is not related to this PR, just redundant code that cause some warnings in Build Result.

File: Microsoft.Dynamics365.UIAutomation.Browser/Logs/StringFormater.cs [hunk 2]
```diff
@@ -0,0 +1,76 @@
+﻿using System;
```
Comment (AngelRodriguez8008) [inline Microsoft.Dynamics365.UIAutomation.Browser/Logs/StringFormater.cs:1]: This two classes can be useful also in project, to easily create more expressive Traces.

File: Microsoft.Dynamics365.UIAutomation.Browser/Logs/TracingService.cs [hunk 3]
```diff
@@ -0,0 +1,129 @@
+﻿using System;
```
Comment (AngelRodriguez8008) [inline Microsoft.Dynamics365.UIAutomation.Browser/Logs/TracingService.cs:1]: This two classes can be useful also in project, to easily create more expressive Traces.

File: Microsoft.Dynamics365.UIAutomation.Sample/ConsoleTraceListener.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
+﻿using System;
```
Comment (AngelRodriguez8008) [inline Microsoft.Dynamics365.UIAutomation.Sample/ConsoleTraceListener.cs:1]: Workaround for MSTest Framework.

File: Microsoft.Dynamics365.UIAutomation.Sample/UCI/ApiTests/TestsBase_Tests.cs [hunk 5]
```diff
@@ -47,5 +45,25 @@ public void NavigationFail_BrowserIsClosed()
                 Assert.IsNull(_client, "Browser still open");
             }
         }
+
```
Comment (AngelRodriguez8008) [inline Microsoft.Dynamics365.UIAutomation.Sample/UCI/ApiTests/TestsBase_Tests.cs:48]: Sample & Tests. Is two times the same test code, but are require two, to reproduce the error. I have no better Idea as mark the test as inconclusive & check visually, that the printed traces are the same. Maybe somebody else has a better idea.

File: Microsoft.Dynamics365.UIAutomation.Sample/UCI/TestsBase.cs [hunk 6]
```diff
@@ -1,8 +1,14 @@
-﻿using System;
+﻿// Enable System.Diagnostics Trace/Debug
+#define DEBUG 
+#define TRACE
+
```
Comment (AngelRodriguez8008) [inline Microsoft.Dynamics365.UIAutomation.Sample/UCI/TestsBase.cs:4]: Require to enable traces, can be anywhere in the assembly. See [documentation](https://docs.microsoft.com/en-us/dotnet/framework/debug-trace-profile/tracing-and-instrumenting-applications) for more infos

File: Microsoft.Dynamics365.UIAutomation.Sample/UCI/TestsBase.cs [hunk 7]
```diff
@@ -12,14 +18,25 @@ public class TestsBase
...
         protected WebClient _client;
       
+        public TestContext TestContext { get; set; } // VSTest fulfill this property before run each test, remove for NUnit
+        public  virtual string CurrentTestName => TestContext.TestName; // NUnit => replace or override this method with: TestContext.CurrentContext.Test.Name
+
+        protected TestsBase()
+        {
+            trace = new TracingService(GetType(), Constants.DefaultTraceSource);
+            trace.Log("Init Tracing Service - Success");
+        }
```
Comment (AngelRodriguez8008) [inline Microsoft.Dynamics365.UIAutomation.Sample/UCI/TestsBase.cs:33]: Some Infos about the current test & init / finish steps

**Review by AngelRodriguez8008 — COMMENTED** [review]
Looks good!

**Review by TYLEROL — APPROVED** [review]

**Pull request merged**